LM-53: implement changes to team controller service dto viewmodel

rgillen-uno opened the pull request:
Changes to Team service/controller and tests for new service functionality.

rgillen-uno committed:
LM-53: Add new GetTeamsByProjectId to controller/service and add initial tests.

rgillen-uno committed:
LM-53: Add some new tests for GetTeamByProjectId

kkuczkowski-unosq commented on AdminCore.Services.Tests/ContractServiceTests.cs:
```diff
@@ -161,32 +182,44 @@ public void TestSaveContractAttemptsUpdateIfContractToBeSavedHasIdOfNotZeroAndCo
         StartDate = new DateTime(2018, 12, 10),
-        EndDate = new DateTime(2019, 1, 10)
+        EndDate = new DateTime(2019, 1, 10),
+        Team = new Team
```
I feel like objects like this should be mocked, `Team` object changes and your tests break.

rgillen-uno replied:
Yeah I agree, I tried to fix the already existing tests while keeping the functionality the same so I didn't make any big changes to them.

kkuczkowski-unosq replied:
Maybe instead, we should consider adding a tech debt task that will fix all these tests.

rgillen-uno committed:
LM-53: Refactor Team service tests to reduce code duplication.

kkuczkowski-unosq commented on AdminCore.Services/ContractService.cs:
```diff
@@ -21,29 +21,28 @@ public ContractService(IDatabaseContext databaseContext, IMapper mapper) : base(
-      var contract = DatabaseContext.ContractRepository.Get(x => x.TeamId == teamId && x.EmployeeId == employeeId, null,
-                                              x => x.Team, x => x.Team.Client);
-      return ReturnContractDto(contract);
+      var contractList = DatabaseContext.ContractRepository.Get(x => x.TeamId == teamId && x.EmployeeId == employeeId, null,
```
Unless the `x` is some kind of standard to specify parameters in lambda expressions, it should be `contract` in this case. That goes for all of these lambda expressions...

rgillen-uno replied:
Is refactoring existing code to have more meaningful variable names maybe outside the scope of this PR?

kkuczkowski-unosq replied:
If you are refactoring this function already I would say this change is acceptable, otherwise it will never get fixed.

rgillen-uno replied:
Refactored lambda function variable names.

kkuczkowski-unosq commented on AdminCore.Services/Mappings/ContractMapperProfile.cs:
```diff
@@ -8,8 +8,10 @@ public class ContractMapperProfile : Profile
   {
     public ContractMapperProfile()
```
any tests for this?

rgillen-uno replied:
I don't think there's any tests for mapping profiles currently.

kkuczkowski-unosq replied:
More reason to add tests

rgillen-uno replied:
Cool I'll add an issue in Jira to add tests for Mapping profiles.

rgillen-uno replied:
Added tests for contract mapper

kkuczkowski-unosq commented on AdminCore.WebApp/Controllers/TeamController.cs:
```diff
@@ -46,6 +47,20 @@ public IActionResult GetTeamById(int id)
+    [HttpGet("getTeamsByProjectId/{projectId}")]
+    [ProducesResponseType(typeof(IList<TeamViewModel>), (int)HttpStatusCode.OK)]
+    [ProducesResponseType(StatusCodes.Status204NoContent)]
+    public IActionResult GetTeamsByProjectById(int projectId)
```
I don't see any tests that test this function

rgillen-uno replied:
AdminCore.Services.Tests/TeamServiceTests.cs https://github.com/UnosquareBelfast/admin-backend/blob/bf2267c6be093020c70e1430073eeb8d8d83e0ee/AdminCore.Services.Tests/TeamServiceTests.cs

rgillen-uno replied:
Ah sorry i misread

rgillen-uno replied:
Added tests

rgillen-uno committed:
Merge branch 'LM-50-modify-services-and-controllers-to-handle-new-project-model' into LM-53-implement-changes-to-team-controller-service-dto-viewmodel

rgillen-uno committed:
LM-55: Add new contract controller/service method and initial tests.

rgillen-uno committed:
LM-53: Remove comments.

rgillen-uno committed:
Merge branch 'LM-50-modify-services-and-controllers-to-handle-new-project-model' into LM-53-implement-changes-to-team-controller-service-dto-viewmodel

kkuczkowski-unosq commented on AdminCore.WebApp/Controllers/TeamController.cs:
```diff
@@ -46,6 +47,20 @@ public IActionResult GetTeamById(int id)
+    public IActionResult GetTeamsByProjectById(int projectId)
+    {
+      var teamList = _teamService.GetByProjectId(projectId);
+      if (teamList == null || !teamList.Any())
```
Hmmm... what if projectId is null, you end up with `teamList` null or empty, then you call `ToString()` on the projectId and 💥.

kkuczkowski-unosq replied:
consider not querying for `teamList()` if projectId is useless

kkuczkowski-unosq replied:
I would like to see these scenarios tested

rgillen-uno replied:
ProjectId is an int, it isn't a nullable type so it won't run into the situation that it is null. If you try and hit that endpoint with an invalid int then the middleware will handle that for you and return a 400 bad request I think.

kkuczkowski-unosq replied:
sounds like something c# specific, can you ensure these scenarios are handled correctly and are tested

kkuczkowski-unosq replied:
interesting, this sounds like it's c# specific

rgillen-uno committed:
LM-55: Fix initial test for contract service.

rgillen-uno committed:
LM-53: Add tests for new controller method.

rgillen-uno committed:
LM-55: Add ContractController test for GetContractByProjectId.

rgillen-uno committed:
Merge branch 'LM-50-modify-services-and-controllers-to-handle-new-project-model' into LM-55-implement-changes-to-contract-controller-service-dto-viewmodel

rgillen-uno committed:
LM-53: Refactor tests to reduce code duplication.

rgillen-uno committed:
LM-55: Fix mapper references.

rgillen-uno committed:
LM-53: Change two tests to use mocked mappers.

rgillen-uno committed:
LM-53: Fix team controller test mapper for test to return correct instance.

rgillen-uno committed:
Merge branch 'LM-53-implement-changes-to-team-controller-service-dto-viewmodel' into LM-55-implement-changes-to-contract-controller-service-dto-viewmodel

rgillen-uno committed:
LM-55: Add correct ProducesResponseType attributes to ContractController.

rgillen-uno committed:
LM-53: Update TeamController tests to match format of later PRs.

rgillen-uno committed:
LM-53: Rework how tests ORM and Mappers are used for new Team tests.

rgillen-uno committed:
Merge branch 'LM-53-implement-changes-to-team-controller-service-dto-viewmodel' into LM-55-implement-changes-to-contract-controller-service-dto-viewmodel

rgillen-uno committed:
LM-55: Rework contract controller and service tests.

BrownMark commented on AdminCore.Services.Tests/ClassData/TeamServiceClassData.cs:
```diff
@@ -0,0 +1,27 @@
+    public class TeamServiceClassData
+    {
+        public class adadsdasdasdas : IEnumerable<object[]>
+        {
```
delet this

rgillen-uno replied:
never

BrownMark commented on AdminCore.Services.Tests/TeamServiceTests.cs:
```diff
@@ -0,0 +1,104 @@
+      teamService.GetByProjectId(7);
+
+      // Assert
+      ormContext.Received().TeamRepository.Get(Arg.Any<Expression<Func<Team, bool>>>(),
```
Id specify the numbers of receives.

rgillen-uno replied:
Done

rgillen-uno committed:
Merge branch 'LM-50-modify-services-and-controllers-to-handle-new-project-model' into LM-53-implement-changes-to-team-controller-service-dto-viewmodel

rgillen-uno committed:
LM-53: Update Team Service tests to not mock the repository.

rgillen-uno committed:
Merge branch 'LM-53-implement-changes-to-team-controller-service-dto-viewmodel' into LM-55-implement-changes-to-contract-controller-service-dto-viewmodel

rgillen-uno committed:
LM-55: Modify contract tests to not mock repository.

rgillen-uno committed:
LM-53: Modify mocked db and orm setup for base class.

rgillen-uno committed:
Merge branch 'LM-53-implement-changes-to-team-controller-service-dto-viewmodel' into LM-55-implement-changes-to-contract-controller-service-dto-viewmodel

rgillen-uno committed:
LM-55: Update tests.

BrownMark commented on AdminCore.DAL/Entity Framework/EntityFrameworkRepository.cs:
```diff
@@ -70,10 +70,11 @@ public IQueryable<T> GetAsQueryable(Expression<Func<T, bool>> filter = null, Fun
-      _dbSet.Attach(entityToUpdate);
+      var updatedEntity = _dbSet.Attach(entityToUpdate);
       ((EntityFrameworkContext)_context).Entry(entityToUpdate).State = EntityState.Modified;
+      return updatedEntity.Entity;
```
Might be worth investigating whether or not this causes an additional DB hit.

rgillen-uno committed:
LM-56: implement changes to project controller service dto viewmodel (#88)

* amend

* LM-56: Add project viewmodels and controller.

* LM-56: Modify viewmodels and add db context save statements.

* LM-56: Update project controller endpoints

* LM-56: Update project controller error handling.

* LM-56: Add initial project controller tests.

* LM-56: Update Project controller class data to support GetProjectsById tests.

* LM-56: Add more tests for ProjectController and refactor method name.

* LM-56: Add more ProjectController tests.

* LM-56: Refactor project service and tests.

* LM-56: Update version of NSubstitute in WebApi.Tests and fix project controller tests.

* LM-56: Add and Update project service tests.

* LM-56: Add project mapping tests.

* LM-56: Add correct ProducesResponseType types to ProjectController.

* LM-56: Rework Project Controller/Service unit tests.

* LM-56: Modify project tests to not mock repository.

* LM-56: Update project interface.

* LM-56: Add optional parameters for getProjects and remove redundant methods.

* LM-56: Add more specific spies for Project Service tests.

* LM-56: Modify controller project to use the hand bad request action.

* LM-56: Update project controller tests to use BeEmpty.

rgillen-uno committed:
LM-54: implement changes to dashboard controller service dto viewmodel (#89)

* LM-54: Add dashboard service/controller functionality, tests, mapping and refactor some names.
Add new Dtos and viewmodels.

* LM-54: Add new tests for DashboardService and modify GetTeamDashboardEvents.
- Add tests to handle empty/null list scenarios for GetTeamDashboardEvents of DashboardService
- Modify GetTeamDashboardEvents to handle null returned from ORM.

* LM-54: Add null handling to GetEmployeeTeamSnapshot and tests for dashboard controller.

* LM-54: Add correct ProducesResponseType attributes to DashboardController.

* LM-54: Remove null-conditionals and redundant dashboard service test.

rgillen-uno committed:
LM-53: Specify number of receives in team service tests.

rgillen-uno committed:
Merge branch 'LM-53-implement-changes-to-team-controller-service-dto-viewmodel' of https://github.com/UnosquareBelfast/admin-backend into LM-53-implement-changes-to-team-controller-service-dto-viewmodel
